### ISSUE TITLE
Statement does not exist in ABAPv755(or a parser error), "SELECT"(parser_error)

### ISSUE THREAD

**Issue by smardcode** [abaplint/abaplint#1747]
The following SELECT statement is valid:
```abap
CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
           lc_name2(5) TYPE c VALUE 'name2',
           lc_name3(5) TYPE c VALUE 'name3'.

SELECT vbeln, vbtyp,
 CASE
   WHEN auart = 'ZAMA' THEN @lc_name1
   WHEN auart = 'ZACR' THEN @lc_name2
  ELSE @lc_name3
END AS ernam
FROM vbak
 INTO @DATA(ls_vbak2).
ENDSELECT.
```
![image](https://user-images.githubusercontent.com/60311033/107138418-8d8ca800-6914-11eb-9c5b-9e6c2f41817d.png)

**Comment by larshp** [comment]
https://syntax.abaplint.org/?filter=select#/expression/SQLCase

**Comment by larshp** [comment]
@smardcode try again, 2.66.8

**Comment by smardcode** [comment]
I am again a bit unsure: the previous issue was fixed, but the linter is complaining "Avoid use of ENDSELECT(avoid_use)".
I modified the statement: I added "UP TO 1 ROWS" and "WHERE vbtyp = 'C'". As far as I know, if we do not do a full qualified WHERE condition, we have to do the "SELECT UP TO 1 ROWS ENDSELECT". What do you mean? I can also open another issue, if you want me to.
```abap
CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
           lc_name2(5) TYPE c VALUE 'name2',
           lc_name3(5) TYPE c VALUE 'name3'.

SELECT vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    UP TO 1 ROWS
    INTO @DATA(ls_vbak2)
    WHERE vbtyp = 'C'.
ENDSELECT.

WRITE ls_vbak2.
```

**Comment by larshp** [comment]
I think

```abap
SELECT vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    UP TO 1 ROWS
    INTO @DATA(ls_vbak2)
    WHERE vbtyp = 'C'.
ENDSELECT.
```

can be rewritten as `SELECT SINGLE`, its a matter of code style, typically its a good idea to avoid SELECT ENDSELECT

**Comment by smardcode** [comment]
I have added "ORDER BY" Statement. Then I cannot use "SELECT SINGLE" any more. Still the linter recommends avoiding "ENDSELECT" ... in combination with ORDER BY the "ENDSELECT" is a must.
```abap
CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
           lc_name2(5) TYPE c VALUE 'name2',
           lc_name3(5) TYPE c VALUE 'name3'.

SELECT vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    UP TO 1 ROWS
    INTO @DATA(ls_vbak2)
    WHERE vbtyp = 'C'
    ORDER BY vbeln.
ENDSELECT.

WRITE ls_vbak2.
```

**Comment by g-back** [comment]
You could alternatively just select into table:

```abap
CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
           lc_name2(5) TYPE c VALUE 'name2',
           lc_name3(5) TYPE c VALUE 'name3'.

SELECT vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    UP TO 1 ROWS
    INTO TABLE @DATA(lt_vbak)
    WHERE vbtyp = 'C'
    ORDER BY vbeln.
IF sy-subrc = 0.
  DATA(ls_vbak) = lt_vbak[ 1 ].
ENDIF.
WRITE ls_vbak.
```

**Comment by larshp** [comment]
opening database cursors typically has a performance penalty, using ENDSELECT also creates nesting

and again, it can be rewritten as SELECT SINGLE, however its up to the user to decide their development best practice

```abap
SELECT SINGLE vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    INTO @DATA(ls_vbak)
    WHERE vbtyp = 'C'
    ORDER BY vbeln.
WRITE ls_vbak.
```

**Comment by larshp** [comment]
some developers like to use SELECT SINGLE only when specifying the full primary key, however there is no guarantee, so the next developer reading the code must check primary keys manually every time, so personally I prefer less code to read

**Comment by smardcode** [comment]
> opening database cursors typically has a performance penalty, using ENDSELECT also creates nesting
> 
> and again, it can be rewritten as SELECT SINGLE, however its up to the user to decide their development best practice
> 
> ```abap
> SELECT SINGLE vbeln, vbtyp,
>     CASE
>     WHEN auart = 'ZAMA' THEN @lc_name1
>     WHEN auart = 'ZACR' THEN @lc_name2
>     ELSE @lc_name3
>     END AS ernam
>     FROM vbak
>     INTO @DATA(ls_vbak)
>     WHERE vbtyp = 'C'
>     ORDER BY vbeln.
> WRITE ls_vbak.
> ```

This statement is not possible: If you use "ORDER BY", then you cannot use "SELECT SINGLE".

**Comment by larshp** [comment]
okay, then as @g-back suggested

**Comment by smardcode** [comment]
> You could alternatively just select into a table:
> 
> ```abap
> CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
>            lc_name2(5) TYPE c VALUE 'name2',
>            lc_name3(5) TYPE c VALUE 'name3'.
> 
> SELECT vbeln, vbtyp,
>     CASE
>     WHEN auart = 'ZAMA' THEN @lc_name1
>     WHEN auart = 'ZACR' THEN @lc_name2
>     ELSE @lc_name3
>     END AS ernam
>     FROM vbak
>     UP TO 1 ROWS
>     INTO TABLE @DATA(lt_vbak)
>     WHERE vbtyp = 'C'
>     ORDER BY vbeln.
> IF sy-subrc = 0.
>   DATA(ls_vbak) = lt_vbak[ 1 ].
> ENDIF.
> WRITE ls_vbak.
> ```

Technically it is possible, but I personally don't find it intuitive.

**Comment by larshp** [comment]
no problem, just disable the check in https://rules.abaplint.org/avoid_use/

by default all rules are enabled in the most strict from in abaplint, I dont think I've ever had all rules enabled in a project, its matter of style, taste, architecture, versions etc

**Comment by smardcode** [comment]
```abap
CONSTANTS: lc_name1(5) TYPE c VALUE 'name1',
           lc_name2(5) TYPE c VALUE 'name2',
           lc_name3(5) TYPE c VALUE 'name3'.

SELECT SINGLE vbeln, vbtyp,
    CASE
    WHEN auart = 'ZAMA' THEN @lc_name1
    WHEN auart = 'ZACR' THEN @lc_name2
    ELSE @lc_name3
    END AS ernam
    FROM vbak
    INTO @DATA(ls_vbak)
    WHERE vbtyp = 'C'
    ORDER BY vbeln.
IF sy-subrc = 0.
  WRITE ls_vbak.
ENDIF.
````
This is now a bug, the linter doesn't complain:
![image](https://user-images.githubusercontent.com/60311033/107279884-eb75d880-6a57-11eb-9fdd-9b37498a4a3f.png)

**Comment by larshp** [comment]
yea, thats a bug, suggest opening it as a separate issue? then can se close this one?

**Comment by smardcode** [comment]
Closing here now, as it is fixed. Opened #1751 for a new bug.
